### PR TITLE
BUG: Check for size != 0 before trying to insert #10193

### PR DESCRIPTION
**Pull request by rekcahpassyla**
Proposed fix for #10193

### CODE REVIEW

**Comment by jreback** [comment]
see my comments on the actual issue. This needs to test that it is:

- hitting the ``SetttingWIthCopy`` path
- affecting the referenced Series (e.g. you have a view in this case).

**Comment by rekcahpassyla** [comment]
I had another go- I looked at how `DataFrame` handled the setting of `is_copy` and then raising `SettingWithCopyError` or `SettingWithCopyWarning`, and tried to replicate that logic in `Series`. 

Added tests for the requisite error / warning being raised, and that the original series is unchanged. Had to change a couple of other tests to change the `chained_assignment` option to `None` and back again. 

I haven't squashed commits as it's still WIP for me. Will look again to see if I can make it more consistent with existing functionality, but here's something for now.

**Comment by rekcahpassyla** [comment]
Noted the build failures. Investigating.

File: pandas/tests/test_series.py [hunk 1]
```diff
@@ -1439,6 +1441,30 @@ def test_setitem(self):
         expected = self.series.append(Series([1],index=['foobar']))
         assert_series_equal(s,expected)
 
+        # Test for issue #10193
+        series = pd.TimeSeries()
```
Comment (jreback) [inline pandas/tests/test_series.py]: don't use ``TimeSeries``, just ``Series``

File: pandas/tests/test_series.py [hunk 2]
```diff
@@ -1439,6 +1441,30 @@ def test_setitem(self):
...
+        series[pd.datetime(2012, 1, 1)] = 47
+        expected = pd.TimeSeries(47, [pd.datetime(2012, 1, 1)])
```
Comment (jreback) [inline pandas/tests/test_series.py]: just use ``datetime``

File: pandas/tests/test_series.py [hunk 3]
```diff
@@ -1418,6 +1418,8 @@ def test_slice_float64(self):
         tm.assert_frame_equal(result, expected)
 
     def test_setitem(self):
+        import warnings
+
```
Comment (jreback) [inline pandas/tests/test_series.py]: you don't need this import

File: pandas/tests/test_series.py [hunk 4]
```diff
@@ -4816,8 +4842,10 @@ def test_rank(self):
         tm._skip_if_no_scipy()
         from scipy.stats import rankdata
 
+        pd.set_option('chained_assignment',None)
```
Comment (jreback) [inline pandas/tests/test_series.py]: use the context manager here

File: pandas/tseries/tests/test_period.py [hunk 5]
```diff
@@ -1669,7 +1669,10 @@ def test_index_duplicate_periods(self):
         result = ts[2007]
         expected = ts[1:3]
         assert_series_equal(result, expected)
+        pd.set_option('chained_assignment', None)
         result[:] = 1
+        pd.set_option('chained_assignment', 'raise')
```
Comment (jreback) [inline pandas/tseries/tests/test_period.py]: use the context manager

File: pandas/tests/test_series.py [hunk 6]
```diff
@@ -1439,6 +1441,30 @@ def test_setitem(self):
...
 
+        # Test for issue #10193
+        series = pd.TimeSeries()
+        series[pd.datetime(2012, 1, 1)] = 47
+        expected = pd.TimeSeries(47, [pd.datetime(2012, 1, 1)])
+        assert_series_equal(series, expected)
+
+        series = pd.TimeSeries(0, pd.date_range('2011-01-01', '2011-01-01'))
+        series2 = series[:0]
+
```
Comment (jreback) [inline pandas/tests/test_series.py]: do a couple simliar tests with a Int64Index (e.g. the default index)

**Comment by rekcahpassyla** [comment]
Requested changes from review have been made, also replaced other existing instances of `pd.set_option('chained_assignment', ...)` with context manager, to be consistent.

**Comment by jreback** [comment]
I have to think about this again. This is a 'feature' of numpy. In that if you set something on a view of something else it works, and changes the original, as that is the point. However, in pandas, objects are pure and changing one shouldn't impact another one, so in effect, even though its supported, the view machinery is effectively frowned upon.

@shoyer @jorisvandenbossche thoughts?

**Comment by shoyer** [comment]
@jreback can you clarify the specific behavior you're not sure about?

**Comment by jreback** [comment]
Well its just general view semantics for Series. These almost always apply because these are a single dtyped, while for frames depends on whether its multi-dtype and what you are setting. I am just thinking that this might be confusing.

```
In [1]: s = Series([1,2,3])

In [2]: s
Out[2]: 
0    1
1    2
2    3
dtype: int64

In [3]: s2 = s[1:]

In [4]: s2
Out[4]: 
1    2
2    3
dtype: int64

In [5]: s2.iloc[0] = 5

In [6]: s2
Out[6]: 
1    5
2    3
dtype: int64

In [7]: s
Out[7]: 
0    1
1    5
2    3
dtype: int64
```

**Comment by shoyer** [comment]
I agree, views are a bit confusing, but they're also a fundamental part of the NumPy ecosystem -- and they make a lot of efficient things possible.

I would definitely prefer not to add more SettingWithCopy warnings. To quote @mwaskom from [Twitter](https://twitter.com/michaelwaskom/status/601405003186577408):

> IMO warnings are almost never a good way to communicate usage notes to users for exactly this reason. They annoy people who know what they are doing, and they rarely provide enough context for those who don't.

**Comment by rekcahpassyla** [comment]
For my own edification; is the issue here that setting a value on a Series which is a view should always pass, because "in pandas, objects are pure and changing one shouldn't impact another one", or that it should always fail?

**Comment by jreback** [comment]
closing in favor of #10379

**Pull request closed without merging**